Source/SIMPLib/CoreFilters/ConvertColorToGrayScale.cpp

```diff
@@ -463,118 +414,31 @@ void ConvertColorToGrayScale::execute()
-        serial.convert(0, totalPoints);
-      }
-    }
-    else if(ConversionType::Lightness == convType)
-    {
-#ifdef SIMPL_USE_PARALLEL_ALGORITHMS
-      if(doParallel)
-      {
-        tbb::parallel_for(tbb::blocked_range<size_t>(0, totalPoints), LightnessImpl(inputColorData->getPointer(0), outputGrayData->getPointer(0), comp), tbb::auto_partitioner());
-      }
-      else
-#endif
-      {
-        LightnessImpl serial(inputColorData->getPointer(0), outputGrayData->getPointer(0), comp);
-        serial.convert(0, totalPoints);
-      }
-    }
-    else if(ConversionType::SingleChannel == convType)
-    {
-#ifdef SIMPL_USE_PARALLEL_ALGORITHMS
-      if(doParallel)
-      {
-        tbb::parallel_for(tbb::blocked_range<size_t>(0, totalPoints), SingleChannelImpl(inputColorData->getPointer(0), outputGrayData->getPointer(0), comp, getColorChannel()),
-                          tbb::auto_partitioner());
-      }
-      else
-#endif
-      {
-        SingleChannelImpl serial(inputColorData->getPointer(0), outputGrayData->getPointer(0), comp, getColorChannel());
-        serial.convert(0, totalPoints);
-      }
-    }
-    else
-    {
-
-      setErrorCondition(-62107);
-      ss = QObject::tr("Conversion Algorithm did not match any available type. The type given was %1.").arg(getConversionAlgorithm());
-      notifyErrorMessage(getHumanLabel(), ss, getErrorCondition());
-      return;
-    }
-  }
-
-#if 0
-    size_t totalPoints = m_ImageDataPtr.lock()->getNumberOfTuples();
-
-    float Rfactor = 1.0f;
-    float Gfactor = 1.0f;
-    float Bfactor = 1.0f;
-    if(m_FlattenMethod == SIMPL::ConvertColorToGrayScaleMethod::Average)
-    {
-        Rfactor = 1.0f / 3.0f;
-        Gfactor = 1.0f / 3.0f;
-        Bfactor = 1.0f / 3.0f;
-    }
-    else if(m_FlattenMethod == SIMPL::ConvertColorToGrayScaleMethod::Luminosity)
-    {
-        Rfactor = 0.21f;
-        Gfactor = 0.72f;
-        Bfactor = 0.07f;
-    }
-
 #ifdef SIMPL_USE_PARALLEL_ALGORITHMS
-    tbb::task_scheduler_init init;
-    bool doParallel = true;
+  tbb::task_scheduler_init init;
 #endif
 
-    int32_t comp = m_ImageDataPtr.lock()->getNumberOfComponents();
-
-    //  qDebug() << "ConvertColorToGrayScale: " << m_ConversionFactor << "\n";
-#ifdef SIMPL_USE_PARALLEL_ALGORITHMS
-    if(doParallel == true)
+    switch (convType)
     {
-        tbb::parallel_for(tbb::blocked_range<size_t>(0, totalPoints), ConvertColorToGrayScaleImpl(m_ImageData, m_FlatImageData, Rfactor, Gfactor, Bfactor, comp), tbb::auto_partitioner());
+      case ConversionType::Luminosity:
+        ParallelWrapper::Run<LuminosityImpl>(LuminosityImpl(inputColorData->getPointer(0), outputGrayData->getPointer(0), m_ColorWeights, comp), totalPoints );
+        break;
+      case ConversionType::Average:
+        ParallelWrapper::Run<LuminosityImpl>(LuminosityImpl(inputColorData->getPointer(0), outputGrayData->getPointer(0), {0.3333f, 0.3333f, 0.3333f}, comp) , totalPoints);
```

Comment:
Because of the way floating point numbers are handled, I am still not sold that LuminosityImpl with `{0.3333f, 0.3333f, 0.3333f}` weights is a completely accurate replacement for AverageImpl.  If we are going down this road, I feel using `1 / 3.0f` as a constant would be more accurate than `0.3333f`.  If we were not working with scientific data, I would probably not be as stringent with this.

Reply:
Since it's merged now, I'd leave it.  This isn't a big enough deal to open another PR about.

Reply:
Well I'd rather include that SIMPLib_EXPORT statement in another PR anyway. I'm running it through the unit test now for the SIMPLib and pending that succeeding, I'll submit another PR. I don't disagree about the floating point...err...point though. So I've amended that as well and am testing it. The test cases I wrote for the AverageImpl should be robust enough to throw if there were any rounding errors but I think, at least, the point may be more clear using 1/3.